src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,246 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+
+namespace Grpc.NetCore.HttpClient.Internal
+{
+    internal class GrpcCall<TRequest, TResponse>
+    {
+        private readonly CancellationTokenSource _callCts;
+        private readonly ISystemClock _clock;
+
+        private HttpResponseMessage _httpResponse;
+        private Metadata _trailers;
+
+        public bool Disposed { get; private set; }
+        public bool ResponseFinished { get; set; }
+        public CallOptions Options { get; }
+        public Method<TRequest, TResponse> Method { get; }
+        public Task<HttpResponseMessage> SendTask { get; private set; }
+        public HttpContentClientStreamWriter<TRequest, TResponse> ClientStreamWriter { get; private set; }
+
+        public HttpContextClientStreamReader<TRequest, TResponse> StreamReader { get; private set; }
+
+        public GrpcCall(Method<TRequest, TResponse> method, CallOptions options, ISystemClock clock)
+        {
+            _callCts = new CancellationTokenSource();
+            Method = method;
+            Options = options;
+            _clock = clock;
+        }
+
+        public CancellationToken CancellationToken
+        {
+            get { return _callCts.Token; }
+        }
+
+        public void SendUnary(System.Net.Http.HttpClient client, TRequest request)
```

Comment:
nit: why the "Send" prefix? Does "StartUnary" make more sense (especially for client streaming calls that don't immediately send a request)?

Reply:
It is called Send because these methods call HttpClient.SendAsync.

Client streaming doesn't talk to the server until the first message is sent?